Post-CWG feedback. (#263)
brevzin/cpp_proposals: 2996_reflection/reflection.md

[Special rules concerning reachability apply to synthesized points ([module.reach]{.sref}).]{.note13}

> No member of an injected declaration shall have a name reserved by the implementation; no diagnostic is required.

Reviewer: Probably good enough. Add a cross-ref to lex.name?

Reviewer: I'll just add it.

> An injected declaration whose name is an identifier reserved by the implementation is ill-formed; no diagnostic is required.

Reviewer: This is not correct, we're not talking about the name of the _injected declaration_. We need to talk about the names of the data members of the injected declaration.

Author: Good point. Changed to:

> An injected declaration that introduces a name reserved by the implementation is ill-formed; no diagnostic is required.

Which I think works.

Reviewer: I don't think that's right? Does `struct C { int x; };` introduce `x`?

Author: Trying again.

template <typename T>
auto fn2() {
  static_assert(^^T != ^^int);
    // static assertion failed during instantiation of fn<int>
  return 0;
}

constexpr bool r2 = can_substitute(^^fn2, {^^int});
  // error: instantiation of body of fn<int> is needed to deduce return type
```
:::

[#]{.pnum}

::: example
```cpp
consteval info to_integral_constant(unsigned i) {
  return substitute(^^integral_constant, {^^unsigned, reflect_value(i)});

>     // OK, represents integral_constant<unsigned, i>

Reviewer: I think better to add a variable that's initialized with `to_integral_constant(2)` and say that it represents the type `integral_constant<unsigned, 2>`

Author: Great idea, done.

  - [#.#.#]{.pnum} `@$N$~$K$~@ == u8"_"` is `true`. [Every provided identifier is unique or `"_"`.]{.note}

[#]{.pnum} *Effects*:
Produces an injected declaration `$D$` ([expr.const]) that defines `$C$` and has properties as follows:

- [#.1]{.pnum} The target scope of `$D$` is the scope to which `$C$` belongs ([basic.scope.scope]).
- [#.#]{.pnum} The locus of `$D$` follows immediately after the core constant expression currently under evaluation.
- [#.#]{.pnum} If `$C$` is a specialization of a templated class `$T$`, and `$C$` is not a local class, then `$D$` is an explicit specialization of `$T$`.

> - [#.#]{.pnum} For each `@$r$~$K$~@`, there is a corresponding entity `@$M$~$K$~@` belonging to the class scope of `$D$` with the following properties:

Reviewer: Is an unnamed bit-field an entity?

Author: Yeah, "bit-fields" are included in the list of entities. In terms of why "entity", I liked that it avoided saying "public non-static data member or unnamed bit-field" twice.

Generally tried to clean things up, and to make it more air-tight (e.g., clarifying dumb things like, "Oh, but you didn't say that it's _not_ a bit-field otherwise").

::: std
::: addu

> [1]{.pnum} Subclause [meta.reflection.unary] specifies consteval functions to query the properties of a type during translation.

Reviewer: ```suggestion
[1]{.pnum} Subclause [meta.reflection.unary] specifies consteval functions to query the properties of a type at compile time.
```
This is Library wording, Library talks about compile time, we're not changing this one.

::: std
::: addu

> [1]{.pnum} The consteval functions specified in this subclause query relationships between types during translation.

Reviewer: Ditto:
```suggestion
[1]{.pnum} The consteval functions specified in this subclause query relationships between types at compile time.
```

::: std
```diff
 [...]
  __cpp_impl_raw_strings            200710L
  __cpp_impl_ref_qualifiers         200710L
+ __cpp_impl_reflection             2025XXL
  __cpp_impl_return_type_deduction  201304L
  __cpp_impl_rvalue_references      200610L

>  [...]

Reviewer: This change is unnecessary, don't need more context.

Author: Removed the ones around it.

Reviewer: I mean, just revert back to what it was before. Having like `impl_coroutine` there was useful to demonstrate that we do this stuff. I just said we don't need _more_ context (not _any_).

```
:::

and [version.syn]{.sref}:

::: std
```diff
  [...]
  #define __cpp_lib_reference_from_temporary  202202L // freestanding, also in <type_traits>
  #define __cpp_lib_reference_wrapper         202403L // freestanding, also in <functional>
+ #define __cpp_lib_reflection                2025XXL // also in <meta>
  #define __cpp_lib_remove_cvref              201711L // freestanding, also in <type_traits>
  #define __cpp_lib_result_of_sfinae          201210L // freestanding, also in <functional>, <type_traits>

>   [...]

Reviewer: This change is unnecessary, don't need any context.

Author: Removed the ones around it.
